### PR TITLE
Add Hebrew keyboard

### PR DESCRIPTION
**Pull request by yparitcher**
english - hebrew keyboard first draft, without diacritics, will add them soon.
please review and let me know if i am doing it right, before i spend more time finishing it off.
also how can i change the menu title from en_HE ?
it looks like the hebrew variable name are failing circleci, how can i fix this?
thanks
Reader_2019-Sep-22_133823
Reader_2019-Sep-22_133846
Reader_2019-Sep-22_134321

### CODE REVIEW

**Comment by poire-z** [comment]
Looks fine (but obviously, I can't say anything about the hebrew part - that's 99% :)
But how are you using it as the UI doesn't support RTL? Type a letter and use the cursor keys to move back?! :)

> how can i change the menu title from en_HE ?

I guess you also need to add it here:
https://github.com/koreader/koreader/blob/bcceb94603a6ac934a604f979f35f368e39c094a/frontend/ui/language.lua#L7-L12
But then, it becomes a full UI language, falling back to english I guess?

> it looks like the hebrew variable name are failing circleci, how can i fix this?

I guess Lua is fine with hebrew letters in variable name, as it works and you managed to make some screenshots. So probably a Luacheck issue, not supporting them.
But may be it's saner to keep variable names in pure ascii?

pinging @Frenzie

**Comment by poire-z** [comment]
> how are you using it as the UI doesn't support RTL? Type a letter and use the cursor keys to move back?! :)

In case you want more fun, I guess you could use a trick introduced with the koread keyboard (ko_KR_keyboard.lua), which wraps the inputbox with (very complicated) code. I guess you could use that to wrap `addChar()` so it does add the char, and then, if it detects its a hebrew char, just do an additional `leftChar()` to move the cursor before the added char.

**Comment by yparitcher** [comment]
> But how are you using it as the UI doesn't support RTL? Type a letter and use the cursor keys to move back?! :)

the only thing RTL or LTR is the display,
i just type normally it just displays LTR (backwards) for example:
Reader_2019-Sep-22_142501
Reader_2019-Sep-22_142508


> > how can i change the menu title from en_HE ?
> 
> I guess you also need to add it here:
> 
> https://github.com/koreader/koreader/blob/bcceb94603a6ac934a604f979f35f368e39c094a/frontend/ui/language.lua#L7-L12
> 
> But then, it becomes a full UI language, falling back to english I guess?
> 

it works, i just did not add it to the language menu, like other languages that dont have a translation yet

File: frontend/ui/widget/virtualkeyboard.lua [hunk 1]
```diff
@@ -526,6 +526,7 @@ local VirtualKeyboard = FocusManager:new{
         ja = "ja_keyboard",
         pt_BR = "pt_keyboard",
         ko_KR = "ko_KR_keyboard",
+        en_HE = "en_HE_keyboard",
```
Comment (Frenzie) [inline frontend/ui/widget/virtualkeyboard.lua]: Please keep it alphabetical.

File: frontend/ui/data/keyboardlayouts/en_HE_keyboard.lua [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+local _v_ = en_popup._v_
+local _W_ = en_popup._W_
+local _w_ = en_popup._w_
+local _X_ = en_popup._X_
+local _x_ = en_popup._x_
+local _Y_ = en_popup._Y_
+local _y_ = en_popup._y_
+local _Z_ = en_popup._Z_
+local _z_ = en_popup._z_
+local _א_ = he_popup._א_
```
Comment (Frenzie) [inline frontend/ui/data/keyboardlayouts/en_HE_keyboard.lua]: Is this variable name actually working? luacheck doesn't usually object without reason (see the failing CI at the bottom of the main PR page):

```
Luacheck results
Checking frontend/ui/data/keyboardlayouts/en_HE_keyboard.lua 1 error

    frontend/ui/data/keyboardlayouts/en_HE_keyboard.lua:59:8: expected statement near 'א'

Checking frontend/ui/data/keyboardlayouts/keypopup/he_popup.lua 1 error

    frontend/ui/data/keyboardlayouts/keypopup/he_popup.lua:2:6: expected '}' (to close '{' on line 1) near 'א'
```

Comment (yparitcher) [reply frontend/ui/data/keyboardlayouts/en_HE_keyboard.lua]: the variables work in the emulator, see the screenshots, i can rename all of the in ascii if needed it just will be more work.

**Comment by yparitcher** [comment]
> > also how can i change the menu title from en_HE ?
> 
> Note that `en-HE` would mean Hebrew English, a variety of English like en-US, en-UK, en-AU, en-NZ, etc. I should simply be named `he`.

right now the keyboard is just a extension to the english, replacing cyrillic characters with hebrew ones, so i am not sure if it should be `he` or `en-HE` i don't mind changing it either way.
this is the way i primarily use it: english ui, but also able to type hebrew characters

**Comment by poire-z** [comment]
> i just type normally it just displays LTR (backwards) for example

Oh right :) everything works in the logical order, so you must type them that way (until we really support displaying them inverted - but still keeping them in that logical order for store and search.)

File: frontend/ui/data/keyboardlayouts/en_HE_keyboard.lua (hunk 2 above)
Comment (Frenzie) [reply frontend/ui/data/keyboardlayouts/en_HE_keyboard.lua]: Indeed it seems that it works in LuaJIT, but it doesn't work in plain Lua. We'd better not introduce any needless incompatibility with regular Lua. :-)

**Comment by Frenzie** [comment]
> right now the keyboard is just a extension to the english, replacing cyrillic characters with hebrew ones, so i am not sure if it should be he or en-HE i don't mind changing it either way.

Please keep it as `he` because `en-HE` doesn't make much sense without a special explanation.

**Comment by poire-z** [comment]
> Please keep it as he because en-HE 

But `he` should probably have the hebrew letters on the initial keyboard map (and latin ones on another map) - if it ever gets usable if/when we have RTL support.
This one here is mainly `en` with another map dedicated to hebrew.
We don't have (yet) anything to handle that kind of 2nd level language.
Unless if now that we have the option to choose the keyboard layout via the menu, it becomes less of an issue, and if I ever need to enter japanese, korean or hebrew, I'd just select the keyboard I want. It's just complicated when I want to enter these different letters in a same typing session... as it's not always possible to save what I just typed, go back to menu, select another keyboad, and re-edit what I was typing.

**Comment by Frenzie** [comment]
I want something like holding or swiping on the spacebar to change layouts, probably out of a selection of enabled ones. But like I wrote in the other PR I've been letting it slide because that would take at least a medium time investment, not least of which experimenting with UX/UI, whereas the quick 10-minute menu switch implementation does a lot of good in little time and I should've done it a year ago.

> But he should probably have the hebrew letters on the initial keyboard map (and latin ones on another map) - if it ever gets usable if/when we have RTL support.

Which doesn't change the fact that en-HE is a confusing name. `en-he` might kinda sorta work, especially if you make it something like `en__he`. I'm not entirely sure how clear that would be, but at least it's not plainly saying English of the HE variety.

**Comment by Frenzie** [comment]
@yparitcher Btw, bit of a random question but how did you mark this PR a draft? I don't think I've ever noticed that before.

**Comment by yparitcher** [comment]
should be fixed, `he` for now, has all the letters their variations (via popups), nikkud (punctution), and some (not all) cantillation marks.

**Comment by yparitcher** [comment]
> @yparitcher Btw, bit of a random question but how did you mark this PR a draft? I don't think I've ever noticed that before.

when opening the PR on the submit button there is a down arrow to select a draft PR, for not ready PR's

File: frontend/ui/data/keyboardlayouts/he_keyboard.lua [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+local daled = he_popup.daled
+local hey = he_popup.hey
+local vov = he_popup.vov
+local zayin = he_popup.zayin
+local tes = he_popup.tes
+local yud = he_popup.yud
+local chof = he_popup.chof
+local lamed = he_popup.lamed
+local mem = he_popup.mem
+local mem_sofis = he_popup.mem_sofis
```
Comment (Frenzie) [inline frontend/ui/data/keyboardlayouts/he_keyboard.lua]: Pity we can't keep 'em three characters but clarity over ASCII art. :-)

**Comment by yparitcher** [comment]
i might add the rest of the cantillation marks via popups in a seperate PR, this one is ready for now

**Comment by poire-z** [comment]
> any comments @poire-z ?

Nope, looks fine to me.
No pb having `he` as a global UI language? It won't magically popup on transiflex and have more people disappointed as we don't display what they added nicely?

**Comment by Frenzie** [comment]
@poire-z Remember when you added a hundred languages for Google Translate or some such? All complete translations have contained the word Hebrew for ages. ;-)

**Comment by Frenzie** [comment]
But also the strings in question here aren't translated. Which is pretty useful when testing some stuff in Chinese. ;-)

**Comment by poire-z** [comment]
I didn't mean the strings in this PR.
Just that now, Hebrew is a UI language. So, people may select it, and find ... english.
Then, they'd go to Transiflex translating all our english strings to hebrew, and the next day go download a nightly, where they'll see `detalsnart yeht sgnirts eht lla` like that ! :)

**Comment by Frenzie** [comment]
I decoupled the language names from the menu items when I added the keyboard selection. My goal was merely to access them through `Language:getLanguageName()` but it also happily obviates any such concerns. :-)

**Comment by poire-z** [comment]
Oh, right, the available menu languages are in another section below in language.lua! :+1:

**Pull request merged**